diff --git a/src/map/atcommand.cpp b/src/map/atcommand.cpp
index d3987f8..95bf60e 100644
--- a/src/map/atcommand.cpp
+++ b/src/map/atcommand.cpp
@@ -10644,6 +10644,18 @@ ACMD_FUNC(battleinfo)
 	return 0;
 }
 
+ACMD_FUNC(showrecovery)
+{
+	sd->showrecovery = !sd->showrecovery;
+	pc_setglobalreg(sd, add_str("AT_showrecovery"), sd->showrecovery);
+	if (sd->showrecovery) {
+		clif_displaymessage(fd, "Hp/sp recovery numbers is visible");
+	} else {
+		clif_displaymessage(fd, "Hp/sp recovery numbers is hidden");
+	}
+	return true;
+}
+
 #include "../custom/atcommand.inc"
 
 /**
@@ -10923,6 +10935,7 @@ void atcommand_basecommands(void) {
 		ACMD_DEF(whobuys),
 		ACMD_DEF(whosell),
 		ACMD_DEF(battleinfo),
+		ACMD_DEF(showrecovery),
 		ACMD_DEF(set),
 		ACMD_DEF(undisguiseguild),
 		ACMD_DEF(disguiseguild),
diff --git a/src/map/pc.cpp b/src/map/pc.cpp
index d303951..eb818d3 100755
--- a/src/map/pc.cpp
+++ b/src/map/pc.cpp
@@ -9356,7 +9356,11 @@ int pc_itemheal(struct map_session_data *sd, t_itemid itemid, int hp, int sp)
 			hp = 0;
 	}
 
-	return status_heal(&sd->bl, hp, sp, 1);
+	if (sd->showrecovery) // -
+		return status_heal(&sd->bl, hp, sp, 2);
+	else
+		return status_heal(&sd->bl, hp, sp, 1);
+	
 }
 
 /*==========================================
diff --git a/src/map/pc.hpp b/src/map/pc.hpp
index 6eafda9..28741ed 100644
--- a/src/map/pc.hpp
+++ b/src/map/pc.hpp
@@ -597,6 +597,7 @@ struct map_session_data {
 
 	bool party_creating; // whether the char is requesting party creation
 	bool party_joining; // whether the char is accepting party invitation
+	bool showrecovery;
 	int party_invite, party_invite_account; // for handling party invitation (holds party id and account id)
 	int adopt_invite; // Adoption
 