diff --git a/src/map/clif.cpp b/src/map/clif.cpp
index 9adba65..26a8ed5 100644
--- a/src/map/clif.cpp
+++ b/src/map/clif.cpp
@@ -10224,14 +10224,14 @@ void clif_name( const block_list* src, const block_list* bl, send_target target
 				packet.packet_id = HEADER_ZC_ACK_REQNAMEALL_NPC;
 				packet.gid = bl->id;
 
-				char title_line[NAME_LENGTH] = {};
-				safesnprintf( title_line, sizeof(title_line), "%s (%u%%)", md->name, get_percentage( md->status.hp, md->status.max_hp ) );
-				memcpy( packet.title, title_line, NAME_LENGTH );
-
 				char name_line[NAME_LENGTH] = {};
-				safesnprintf( name_line, sizeof(name_line), "%s %s", get_mob_race_name(md->status.race), get_mob_size_tag(md->status.size) );
+				safesnprintf( name_line, sizeof(name_line), "%s (%u%%)", md->name, get_percentage( md->status.hp, md->status.max_hp ) );
 				safestrncpy( packet.name, name_line, NAME_LENGTH );
 
+				char title_line[NAME_LENGTH] = {};
+				safesnprintf( title_line, sizeof(title_line), "%s %s", get_mob_race_name(md->status.race), get_mob_size_tag(md->status.size) );
+				memcpy( packet.title, title_line, NAME_LENGTH );
+
 				if (md->status.def_ele >= ELE_NEUTRAL && md->status.def_ele < ELE_MAX)
 					packet.groupId = 51 + md->status.def_ele;
 
